drivers/char/rust_example/src/lib.rs

```diff
@@ -36,6 +37,12 @@ impl KernelModule for RustExample {
         println!("Parameters:");
         println!("  my_bool:  {}", my_bool.read());
         println!("  my_i32:   {}", my_i32.read());
+
+        let rust_mutex = Mutex::new(1, "rust_mutex");
+        println!("rust_mutex: {:?}", rust_mutex);
+        let guard = rust_mutex.lock();
+        println!("data: {:?}", guard);
+        drop(guard);
```

Comment:
- Perhaps a block instead of `drop` to show the advantage of RAII?
  - Also a newline after the block, to separate things a bit.
  - `rust_mutex` -> `mutex`?

Reply:
Not sure how you mean the first point. would that not just block this thread?

Reply:
if a do `let guard2 = rust_mutex.lock()` I get a page fault. Not sure if this is becaus my thread already has the lock? or is something else broken?

The page fault is because it does not block, and so returns a MutexGuard. This calls mutex_unlock() to times.

Reply:
> Not sure how you mean the first point. would that not just block this thread?

I meant writing it this instead:

    {
        let guard = rust_mutex.lock();
        println!("data: {:?}", guard);
    }

So that the `drop()` is automatic (RAII). This also has the advantage of showing what exactly is covered by the lock (the `{ ... }` scope).

Reply:
Oh, I see. Will do. 

But we should somehow address the lock error I found. But I don't know how, and if it's by the kernels mutex_lock function design

Reply:
> Not sure if this is becaus my thread already has the lock?

Yeah, recursive locking is not allowed (with a plain `mutex_lock()`). The kernel code assumes you use the functions correctly for performance. But you can confirm it is that by enabling `CONFIG_DEBUG_MUTEXES` -- it should complain that you are trying to lock it twice.

> This calls mutex_unlock() to times.

That means we need to mark some functions in the API as unsafe (to tell people they need to validate themselves they are meeting the requirements), or write the wrapper in a way that makes that case impossible.

Reply:
> if it's by the kernels mutex_lock function design

Take a look at `Doc/locking/mutex-design.rst`.